core/cog-host-map-handler.c

```diff
@@ -0,0 +1,299 @@
+    }
+}
+
+static void
+cog_host_map_handler_dispose(GObject *object)
+{
+    CogHostMapHandler *self = COG_HOST_MAP_HANDLER(object);
+
+    g_hash_table_remove_all(self->map);
+    g_clear_pointer(&self->map, g_hash_table_unref);
+
+    g_clear_object(&self->fallback);
+
+    G_OBJECT_CLASS(cog_host_map_handler_parent_class)->dispose(object);
+}
+
+static void
+cog_host_map_handler_class_init(CogHostMapHandlerClass *klass)
+{
+    GObjectClass *object_class = G_OBJECT_CLASS(klass);
+    object_class->get_property = cog_host_map_handler_get_property;
+    object_class->set_property = cog_host_map_handler_set_property;
+    object_class->dispose = cog_host_map_handler_dispose;
+
+    /**
+     * CogHostMapHandler:fallback-handler:
+     *
+     * Optional #CogRequestHandler used as fallback when no map entries match
+     * the URI path.
+     */
+    s_properties[PROP_FALLBACK_HANDLER] =
+        g_param_spec_object("fallback-handler",
+                            "Fallback handler",
+                            "Handler used as fallback for unhandled map entries",
+                            COG_TYPE_REQUEST_HANDLER,
+                            G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+    g_object_class_install_properties(object_class, N_PROPERTIES, s_properties);
+}
+
+static void
+cog_host_map_handler_init(CogHostMapHandler *self)
+{
+    self->map = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_object_unref);
+}
+
+/**
+ * cog_host_map_handler_new: (constructor)
+ * @fallback_handler: (nullable) (transfer none): a #CogRequestHandler
+ *    to use as fallback
+ *
+ * Create a new handler.
+ *
+ * Returns: (transfer full): A handler with no map entries configured.
+ */
+CogRequestHandler *
+cog_host_map_handler_new(CogRequestHandler *fallback_handler)
+{
+    return g_object_new(COG_TYPE_HOST_MAP_HANDLER, "fallback-handler", fallback_handler, NULL);
+}
+
+/**
+ * cog_host_map_handler_contains:
+ * @self: a #CogHostMapHandler
+ * @host: URI host to match
+ *
+ * Check if there is already mounted a route for the @host.
+ *
+ * Returns: Whether the route already exists.
+ */
+gboolean
+cog_host_map_handler_contains(CogHostMapHandler *self, const char *host)
+{
+    g_return_val_if_fail(COG_IS_HOST_MAP_HANDLER(self), FALSE);
+    g_return_val_if_fail(host != NULL, FALSE);
+
+    if (g_hash_table_contains(self->map, host))
+        return TRUE;
+
+    return FALSE;
+}
+
+/**
+ * cog_host_map_handler_mount:
+ * @self: a #CogHostMapHandler
+ * @host: URI host to match
+ * @handler: (transfer none): a request handler for the matched host
+ *
+ * Adds a route to the handler.
+ *
+ * Configures a route which matches @host in URI, and dispatches
+ * requests to a given @handler.
+ *
+ * Returns: Whether the route was successfully added.
+ */
+gboolean
+cog_host_map_handler_mount(CogHostMapHandler *self, const char *host, CogRequestHandler *handler)
```

Comment:
I would call this `cog_host_map_handler_add()` because the “mount” verb indicates that something is being attached to a location, and this handler is _not_ using the URI “location” (the path) at all, only the *host* part.